src/main/kotlin/com/example/delivery/service/OrderService.kt

```diff
@@ -27,71 +32,96 @@ class OrderService(
+            .flatMap { user ->
+                val requestedProductsIds = createOrderDTO.items.map { it.productId }
+                productRepository.findAllByIds(requestedProductsIds)
+                    .collectList()
+                    .flatMapMany { productsList ->
+                        if (productsList.size == createOrderDTO.items.size) {
+                            Flux.fromIterable(productsList.map { it.toDomain() })
+                        } else {
+                            checkProductAvailability(productsList, requestedProductsIds)
+                        }
+                    }
+                    .flatMap { product -> checkProductAmount(createOrderDTO, product) }
+                    .collectList()
+                    .flatMap { items ->
+                        saveOrder(items, createOrderDTO, user)
+                    }
+            }
+    }
 
-        val products = productRepository.findAllByIds(createOrderDTO.items.map { it.productId }).map { it.toDomain() }
+    fun deleteById(id: String): Mono<Void> {
+        return orderRepository.deleteById(id)
+    }
 
-        if (products.size != createOrderDTO.items.size) {
-            val missingProductId = createOrderDTO.items
-                .map { it.productId }
-                .first { itemId -> !products.map { it.id }.contains(itemId) }
-            throw NotFoundException("Product with id $missingProductId")
-        }
+    fun updateOrder(id: String, updateOrderDTO: UpdateOrderDTO): Mono<DomainOrder> {
+        return orderRepository.updateOrder(id, updateOrderDTO.toUpdate())
+            .map { it.toDomain() }
+            .switchIfEmpty(Mono.error(NotFoundException("Order with id $id doesn't exists")))
+    }
+
+    fun updateOrderStatus(id: String, status: String): Mono<DomainOrder> {
+        return orderRepository.updateOrderStatus(id, MongoOrder.Status.valueOf(status))
+            .map { it.toDomain() }
+            .switchIfEmpty(Mono.error(NotFoundException("Order with id $id doesn't exists")))
+    }
 
-        createOrderDTO.items.forEach { orderItem ->
-            val product = products.first { orderItem.productId == it.id }
-            if (orderItem.amount > product.amountAvailable) {
-                throw ProductAmountException(
+    fun getAllByUserId(id: String): Flux<DomainOrder> {
+        return orderRepository.findAllByUserId(id).map { it.toDomain() }
+    }
+
+    private fun checkProductAvailability(products: List<MongoProduct>, itemsIds: List<String>): Flux<DomainProduct> {
+        val productIds = products.map { it.id.toString() }
+        val missingProductId = itemsIds.first { it !in productIds }
+        return Flux.error(NotFoundException("Product with id $missingProductId doesn't exist"))
+    }
+
+    private fun checkProductAmount(
+        createOrderDTO: CreateOrderDTO,
+        product: DomainProduct,
+    ): Mono<MongoOrder.MongoOrderItem> {
```

Comment:
do you perform any blocking operations inside this method?

Reply:
I think no 🤔